Nome: Ramon Gomes Paulino   RA:250504

In [9]:
import numpy as np
import matplotlib.pyplot as plt

Questão 1)

In [10]:
#Dados da questão → Ax = b
A = np.array([[0.1, 0.2, 0.3],
              [0.4, 0.5, 0.6],
              [0.7, 0.8, 0.9]
             ])

b = np.array([0.1, 0.3, 0.5])

##### (a)Mostre que a matriz A é singular e descreva o conjunto de soluções do sistema linear Ax = b.

A matriz A é singular se o seu determinante for igual a zero. logo:

det(A) = (0.1)*(0.5)*(0.9) + (0.4)*(0.8)*(0.3) + (0.7)*(0.2)*(0.6) - (0.3)*(0.5)*(0.7) - (0.2)*(0.4)*(0.9) - (0.1)*(0.6)*(0.8)

que resulta em:

0.045 + 0.096 + 0.084 - 0.105 - 0.072 - 0.048 = 0.0

logo, temos uma matriz singular.

Para encontrarmos o conjunto solução, podemos utilizar da matriz extendida:

[[0.1, 0.2, 0.3, 0.1],                                 
 [0.4, 0.5, 0.6, 0.3], se fizermos que L2 = L2-4*L1 <br>
 [0.7, 0.8, 0.9, 0.5] <br>
]

[[0.1,  0.2,  0.3,  0.1], <br>
 [0.0, -0.3, -0.6, -0.1], <br>
 [0.7,  0.8,  0.9,  0.5], se fizermos que L3 = L3-7*L1 <br>
]

[[0.1,  0.2,  0.3,  0.1], <br> 
 [0.0, -0.3, -0.6, -0.1], <br>
 [0.0, -0.6, -1.2, -0.2], se fizermos que L3 = L3-2*L2 <br>
]

[[0.1,  0.2,  0.3,  0.1], <br>
 [0.0, -0.3, -0.6, -0.1], <br>
 [0.0,  0.0,  0.0,  0.0], <br>
]

com isto, temos o sistema linear:

{    0.1x + 0.2y + 0.3z =  0.1  
{         - 0.3y - 0.6z = -0.1  → y = -(0.6z-0.1)/0.3  → y = -2z + 1/3 (ii) <br>
{                     0 = 0     → Z = Z <br> 

Disto, temos a seguinte solução:

{    0.1x + 0.2y + 0.3z =  0.1  → trazendo (ii) para cá → 0.1x + 0.2(-2z + 1/3) + 0.3z = 0.1   → x-(z-0.66666) = 1 → x = z + 0.3333 <br>
{         - 0.3y - 0.6z = -0.1  <br>
{                     0 = 0    

Logo, a solução é da forma:<br>
x =   z + 0.3333 <br>
y = -2z + 0.3333<br>
z =   z<br>

-------------------------------------------------------------------------------

##### (b) Se fôssemos usar o método da eliminação de Gauss com pivotamento parcial para resolver este sistema usando aritmética exata, em que ponto o processo falha?

Ocorreria falha pois o método de eliminação de Gauss com pivoteamento parcial teria muitos erros de arredondamento, devido aos denominadores menores que 1, portanto, poderia se propagar erros durante o procedimento de maneira a perder o valor exato.

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

##### (c) Como algumas das entradas da matriz A não são exatamente representáveis em um sistema de ponto flutuante na base 2, a matriz não é singular quando inserida num computador. Portanto, o método da eliminação de Gauss não falha ao tentar resolver o sistema linear. Resolva o sistema linear Ax = b em python usando a biblioteca Numpy ou SciPy.. Compare a solução calculada com sua descrição do conjunto solução definida no item (a).

In [11]:
det_A = np.linalg.det(A)
print("Determinante de A:", det_A)

if abs(det_A) == 0:
    print("A matriz A é singular.")
else:
    print("A matriz A não é singular.")

#para encontrar seus valores, podemos usar a função np.linalg.solve, mas como a matriz é singular, o sistema pode ter infinitas soluções ou nenhuma solução.
#Vamos verificar se o sistema tem soluções:
try:
    solucao = np.linalg.solve(A, b)
    print("\n Solução única encontrada:\n",solucao)
except np.linalg.LinAlgError:
    print("O sistema não tem solução única.\n")

Determinante de A: 6.661338147750926e-18
A matriz A não é singular.

 Solução única encontrada:
 [ 0.16145833  0.67708333 -0.171875  ]


--<br>

Obtivemos como determinante um valor muito próximo de 0, porém, o computador não entende como 0 dado os arredondamentos, por isso, não considera que a matriz seja singular.
Pelo mesmo motivo, a solução calculada do sistema é única, mas como vimos no item (a) a solução tem a forma: 

x =   z + 0.3333 <br>
y = -2z + 0.3333<br>
z =   z<br>

E se aplicarmos o valor de z calculado como único pelo computador, teremos os mesmos resultados:

x =   -0.171875 + 0.3333   =  0.16145833<br>
y = -2(-0.171875) + 0.3333 =  0.67708333 <br>
z =   -0.171875

----------------------------------------------------

(d) Qual é o número de condição da matriz A? Quantos dígitos de precisão você espera na solução numé-
rica do sistema linear?

O número de condição de uma matriz é dada por k(A) = ||A^-1||*||A||

Como a matriz é singular, não tem inversa. Com isto, por definição, temos uma matriz que contem número de condição infinito. Como podemos ver abaixo, o código nos dá um valor extremamente alto, indicando o infinito, dado que computacionalmente esse sistema não é singular.

In [12]:
cond_A = np.linalg.cond(A)
print("Número de condição da matriz A:", cond_A)

Número de condição da matriz A: 2.1118968335779856e+16


In [13]:
vetor_residual  = b -np.dot(A, solucao)
print("\n Vetor residual:\n", vetor_residual)


 Vetor residual:
 [1.38777878e-17 5.55111512e-17 0.00000000e+00]


Logo, podemos esperar 16 digitos de precisão em x e y, dado seu erro residual na 17 casa depois da vírgula. Já em z, temos uma precisão indefinida.

------------

(2) (a) Determine a matriz A e o vetor b do sistema linear Ax = b cuja solução corresponde ao vetor f =
[f1, . . . , f13] contendo as forças nas barras da treliça.

In [14]:
al = np.sqrt(2)/2

A2 = np.array([[ 0,  1,  0,  0,   0, -1,  0,  0,   0,  0,  0,   0,  0],
               [ 0,  0,  1,  0,   0,  0,  0,  0,   0,  0,  0,   0,  0],
               [al,  0,  0, -1, -al,  0,  0,  0,   0,  0,  0,   0,  0],
               [al,  0,  1,  0,  al,  0,  0,  0,   0,  0,  0,   0,  0],
               [ 0,  0,  0,  1,   0,  0,  0, -1,   0,  0,  0,   0,  0],
               [ 0,  0,  0,  0,   0,  0,  1,  0,   0,  0,  0,   0,  0],
               [ 0,  0,  0,  0,  al,  1,  0,  0, -al, -1,  0,   0,  0],
               [ 0,  0,  0,  0,  al,  0,  1,  0,  al,  0,  0,   0,  0],
               [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  1,  0,   0, -1],
               [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  0,  1,   0,  0],
               [ 0,  0,  0,  0,   0,  0,  0,  1,  al,  0,  0, -al,  0],
               [ 0,  0,  0,  0,   0,  0,  0,  0,  al,  0,  1,  al,  0],
               [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  0,  0,  al,  1]])

b2 = np.array([0,10,0,0,0,0,0,15,0,20,0,0,0])

----

(b) Resolva o sistema linear usando o python usando a biblioteca Numpy ou SciPy.

In [15]:
np.linalg.matrix_rank(A2)
print("\n Posto da matriz A2:", np.linalg.matrix_rank(A2))
f = np.linalg.solve(A2, b2).tolist()
print("\n Solução do sistema Ax = b (forças nas barras da treliça):\n", f)


 Posto da matriz A2: 13

 Solução do sistema Ax = b (forças nas barras da treliça):
 [-28.2842712474619, 20.0, 10.0, -29.999999999999996, 14.142135623730947, 20.0, 0.0, -29.999999999999996, 7.071067811865476, 24.999999999999996, 20.0, -35.35533905932737, 24.999999999999996]


---

(c) Observe que a matriz A do sistema linear é esparsa. É possível resolver o sistema linear usando os
métodos de Gauss-Jacobi? E usando o método de Gauss-Seidel, é possível resolver o sistema linear
Ax = b? Justifique sua resposta.

É possível resolver o sistema usando o método de Gauss-Jacobi, pois fazendo uma manipulação de linhas, podemos ver que os valores da diagonal principal são diferentes de 0.

```
 [al,  0,  1,  0,  al,  0,  0,  0,   0,  0,  0,   0,  0]
 [ 0,  1,  0,  0,   0, -1,  0,  0,   0,  0,  0,   0,  0]
 [ 0,  0,  1,  0,   0,  0,  0,  0,   0,  0,  0,   0,  0]
 [al,  0,  0, -1, -al,  0,  0,  0,   0,  0,  0,   0,  0]
 [ 0,  0,  0,  0,  al,  0,  1,  0,  al,  0,  0,   0,  0]
 [ 0,  0,  0,  0,  al,  1,  0,  0, -al, -1,  0,   0,  0]
 [ 0,  0,  0,  0,   0,  0,  1,  0,   0,  0,  0,   0,  0]
 [ 0,  0,  0,  1,   0,  0,  0, -1,   0,  0,  0,   0,  0]
 [ 0,  0,  0,  0,   0,  0,  0,  1,  al,  0,  0, -al,  0]
 [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  1,  0,   0, -1]
 [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  0,  1,   0,  0]
 [ 0,  0,  0,  0,   0,  0,  0,  0,  al,  0,  1,  al,  0]
 [ 0,  0,  0,  0,   0,  0,  0,  0,   0,  0,  0,  al,  1]

```

É possível resolver o sistema usando o método de Gauss-Seidel, pois ele tem pelo menos uma solução, dado que é possível resolver o sistema, pois temos 13 equações e 13 incógnitas, com isto, temos equações suficientes para encontrar as 13 incógnitas. Também não temos dependência entre as equações (LI).  